inductor: separate bias from PackeLinear for better performance

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93837
* __->__ #93348
* #93341
* #93336


For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor: sepreate bias from PackeLinear for better performance

[ghstack-poisoned]

jgong5 commented on torch/_inductor/lowering.py:
```diff
@@ -964,9 +964,12 @@ def mkl_packed_linear(
-                return TensorBox.create(
-                    ir.MKLPackedLinear.create(x, packed_w, orig_w, b, batch_size)
+                result = TensorBox.create(
+                    ir.MKLPackedLinear.create(x, packed_w, orig_w, None, batch_size)
```
If bias is always not needed, why do we bother pass `None`? How about removing `bias` arg?

XiaobingSuper replied:
Yes, removed it.

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: separate bias from PackeLinear for better performance"



For PakedLinear with has bias, we always copy bias to output before doing the computation:
https://github.com/pytorch/pytorch/blob/d7a3f2128fb4457dd60fd5d23e77d2c66a8b0f02/aten/src/ATen/native/mkldnn/Linear.cpp#L389-L397.

This PR separates bias from it which can make the bias add fused with the post-op.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
